src/Illuminate/Support/Url.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Illuminate\Support;
+
+use Illuminate\Contracts\Support\Arrayable;
+
+class Url implements Arrayable
+{
+    /**
+     * Constructor.
+     */
+    public function __construct(
+        public ?string $scheme = null,
+        public ?string $host = null,
+        public ?string $port = null,
+        public ?string $user = null,
+        public ?string $pass = null,
+        public ?string $path = null,
+        public ?string $query = null,
+        public ?string $fragment = null,
+    ) {
+    }
+
+    /**
+     * Parse a URL string into a URL object.
+     *
+     * @param  string  $url
+     * @return static
+     */
+    public static function parse(string $url): static
+    {
+        $components = parse_url($url) ?? [];
+
+        return new static(
+            $components['scheme'] ?? null,
+            $components['host'] ?? null,
+            $components['port'] ?? null,
+            $components['user'] ?? null,
+            $components['pass'] ?? null,
+            $components['path'] ?? null,
+            $components['query'] ?? null,
+            $components['fragment'] ?? null,
+        );
+    }
+
+    /**
+     * Get the URL query parameters.
+     *
+     * @return \Illuminate\Support\UrlQueryParameters
+     */
+    public function query(): UrlQueryParameters
+    {
+        return UrlQueryParameters::parse($this->query);
```

Comment:
```suggestion
    public function query($item = null): UrlQueryParameters
    {
        $this->parsedQuery ??= UrlQueryParameters::parse($this->query);
        
        return $item ? $this->parsedQuery->get($item) : $this->parsedQuery;
```

Reply:
Thanks for the review!

Not sure if we want to cache the parsed query here... Since the `Url` properties are public, they can be changed at any time which wouldn't reflect the true state of the object (I know this is edge case though).

Reply:
I agree that I would expect that I can pull out the query parameter object, make changes, and have it reflected on the URL.

I also agree that the public properties make that hard.

Perhaps the `query` property should be a instance of the `UrlQueryParameters`, rather than a string.

Reply:
This has been completed! 👍